### ISSUE TITLE
JSON deserialization of grain states does not bind GrainReferences to runtime

### ISSUE THREAD

**Issue by heikki-heikkinen-ptcs** [OrleansContrib/Orleans.Redis#10]
This provider cannot be used as PubSub store with JSON serialization format because the `PubSubRendezvousGrain`'s `PubSubGrainState` contains GrainReferences on in the Consumers set (`PubSubSubscriptionState.consumerReference`) which does not get bound to the runtime when the state is deserialized with `JsonConvert` in `RedisGrainStorage.ReadStateAsync`:
```
if (_options.UseJson)
{
    grainState.State = JsonConvert.DeserializeObject(valueEntry.Value, grainState.State.GetType(), _jsonSettings);
}
```

The `_jsonSettings` does not contain a converter that would bind `IAddressable` instances to the runtime.

This same issue has been fixed to this library:
commit: https://github.com/OrleansContrib/Orleans.Providers.MongoDB/commit/9cc966597a11c6fdd4c1103ebf8b9a1c93547429
issue: https://github.com/OrleansContrib/Orleans.Providers.MongoDB/issues/21

Also upgrading from Orleans.Persistence.Redis from 3.0.1 to 3.1 changes the default serialization format in `RedisStorageOptions` to JSON.
```
/// <summary>
/// Whether or not to use JSON for serialization.
/// </summary>
public bool UseJson { get; set; } = true;
```

This is a change in default behavior which I don't think should have been done in a minor version bump. This is a breaking change for usages that do not explicitly set the `UseJson` property in `AddRedisGrainStorage`.

PS. Would it be possible to add releases to this GitHub page with changelogs. Otherwise we'll have to read every line on code in the disassembled NuGet package before upgrading the library.

**Comment by ReubenBond** [comment]
uhoh, thanks for pointing this out @heikki-heikkinen-ptcs. Are you able to submit a PR? We can make a quick release afterwards.

**Comment by heikki-heikkinen-ptcs** [comment]
@ReubenBond Sure, I'll cook this up on my pipe vision breaks. BTW are there any agreed practices for fixing code style, beautification and unrelated issues? There are some that are not directly related.
Examples:
- `private const string _writeScript`: This might be a naming convention issue depending on what they are. Commonly const fields are in PascalCase.
- Structured logging issue:
``` 
var initMsg = string.Format("Init: Name={0} ServiceId={1} DatabaseNumber={2} UseJson={3} DeleteOnClear={4}",
                            _name, _serviceId, _options.DatabaseNumber, _options.UseJson, _options.DeleteOnClear);
_logger.LogDebug($"RedisGrainStorage {_name} is initializing: {initMsg}"); 
```
In this case there are several issues:
 - _logger.LogDebug: call takes an interpolated string as a parameter. This should be exploded into a structured logging call where the format arguments are passed to the LogDebug call as parameters
 - The initMsg should not be passed as a single parameter but should be incorporated as part of the entire log message using structured logging.
 - If there is a good reason for adding the initMsg as a single log parameter the string should probably be formed wwith `FormattableString.Invariant($"message goes here as interpolated string")`

- Uninformative exceptions:
```
                    var etagEntry = hashEntries.Single(e => e.Name == "etag");
                    var valueEntry = hashEntries.Single(e => e.Name == "data");
```
- Sure the `Single` extension method is handy but it results in an uninformative InvalidOperationException with a generic message "Sequence contains more than one matching element" which doesn't tell you anything about the context.

Should I tidy things up or just do the bare minimum here?

**Comment by heikki-heikkinen-ptcs** [comment]
Pushed the rename to global.json to the PR by mistake. It was only preventing the solution from loading as far as I know if you don't have that exact version of dotnet SDK installed.
https://github.com/OrleansContrib/Orleans.Redis/pull/11

**Comment by ReubenBond** [comment]
Fixed by #11
